### ISSUE TITLE
Contrib AdaMax implementation producing NaNs on GPU

### ISSUE THREAD

**Issue by benleetownsend** [tensorflow/tensorflow#26256]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu
- TensorFlow installed from (source or binary): Source
- TensorFlow version (use command below): Tested on 1.12 and 1.13
- Python version: 3.6
- CUDA/cuDNN version: 9
- GPU model and memory: verified on 1080ti and titan v

**Describe the current behavior**

On GPU AdaMax from `tf.contrib.opt.AdaMaxOptimizer` appears to apply NaNs to variables. Seems fine on CPU and bizarrely if any ops are put as a control_dependency to the apply_grads call then everything seems fine.

**Describe the expected behavior**
Not to produce NaNs.

**Code to reproduce the issue**
```
import tensorflow as tf

a = tf.get_variable("a", shape=[10000])
b = tf.get_variable("b", shape=[10000])
loss = tf.minimum(tf.reduce_mean(a), tf.reduce_mean(b))
sess = tf.Session()

print("===== NORMALLY ======")

opt_op= tf.contrib.opt.AdaMaxOptimizer().minimize(loss)
sess.run(tf.global_variables_initializer())
for a in range(10):
    print(sess.run([opt_op, loss])[1])

print("===== WITH NOOP ======")

opt = tf.contrib.opt.AdaMaxOptimizer()
grads_and_vars = opt.compute_gradients(loss)
noop = tf.no_op()
    
with tf.control_dependencies([noop]):
    opt_op_fixed = opt.apply_gradients(grads_and_vars)

sess.run(tf.global_variables_initializer())
for a in range(10):
    print(sess.run([opt_op, loss])[1])
```

The above code produces on my machine:
```
===== NORMALLY ======
-3.637023e-05
-0.00037572667
nan
nan
nan
nan
nan
nan
nan
nan
===== WITH NOOP ======
-9.1626454e-05
-0.00018933268
-0.00028703889
-0.00038474516
-0.0004824514
-0.0005801577
-0.0006778639
-0.00077557017
-0.0008732763
-0.00097098254
```

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by madisonmay** [comment]
I can reproduce on GPU device on tf 1.11.0 as well.

**Comment by facaiy** [comment]
Thank you for pointing it out. I'll take a look at it.

**Comment by facaiy** [comment]
Hi, could you take a try https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adamax ?  Because tf.contrib has been deprecated, and tf.contrib.Adamax will be deleted (replaced by tf.keras.AdaMax). Thank you :-)

**Comment by maziello** [comment]
In TF 2.0, tf.keras.optimizers.Adamax (and I think Adam as well) has the same problem. I had to use a different optimizer (SDG) in order to get mobilenet_v2 to train and not produce a NaN loss

**Comment by jvishnuvardhan** [comment]
@benleetownsend I could not reproduce this with TF1.12.0 (gist is [here](https://colab.sandbox.google.com/gist/jvishnuvardhan/aee69fe92d830ba63f1f8b9b706d28f8/untitled118.ipynb). But TF1.13.1 still has the issue (gist is [here](https://colab.sandbox.google.com/gist/jvishnuvardhan/a38562b050931db33b09bddc34451ded/untitled117.ipynb). Both the gist are creating using Google colab with GPU. Thanks!

**Comment by facaiy** [comment]
@jvishnuvardhan Vishnuvardhan, as @maziello reported, keras Adamax has the same problem, could you take a try to reproduce it? I think we should address it in keras optimizer module, rather than contrib module.

**Comment by arthurflor23** [comment]
Here, with tensorflow 2 (tf.keras):
* Adamax: CPU is ok and GPU NaN
* Adam: CPU is ok and GPU is ok

**Comment by daniesso** [comment]
Also have this problem. The other keras optimizers work just fine, Adamax works on CPU but not on GPU, using tensorflow 2.

**Comment by facaiy** [comment]
@tanzhenyu @case540 Can you take a look? Thanks.

**Comment by tanzhenyu** [comment]
Can you try Tensorflow 1.14 with tf.keras.optimizers.Adamax?

**Comment by daniesso** [comment]
Nan loss there as well. If it helps, I'm using a network with an embedding layer and LSTM (CuDNNLSTM), so I assume it could be a problem with either the dense or the sparse update.

**Comment by tanzhenyu** [comment]
Hmm...does the same thing happen on CPU as well?

**Comment by daniesso** [comment]
Nope, Adamax works just fine on the CPU. To elaborate: it happens immediately* and every time with Adamax on the GPU, and never with Adamax on the CPU.

\* On the second batch, after updating the weights once, the loss is nan.

**Comment by Edwin-Koh1** [comment]
Hello, I have this issue on GPU AWS 1.14.0 recently which did not occur prior to the update when they were still using 1.13.1

**Comment by tanzhenyu** [comment]
1.13.1 probably doesn't have GPU kernel that's why it got relocated to CPU.
The GPU kernel was probably introduced after 1.14, which had a major bug.
This is fixed on a 09/11 code change and should be reflected in either 1.15 rc3, or 2.0.
Please try that and let us know if that fixed your issue.

**Comment by Edwin-Koh1** [comment]
I tried it on 2.0 on my local machine and it works on GPU. However, AWS is still stuck on 1.14.0 for now which sucks and the bug is still there.

**Comment by Edwin-Koh1** [comment]
Yes, the NAN's are gone for Adamax optimizer on GPU training on tensorflow 2.0.0 but still persists on 1.14

**Comment by tanzhenyu** [comment]
Ok. I'm gonna close this for now. AWS could update to 1.15 later since that should get fixed.

**Comment by SirRob1997** [comment]
Still having this problem on `1.15` is there any solution besides updating to `2.0.0`?

**Comment by tanzhenyu** [comment]
> Still having this problem on `1.15` is there any solution besides updating to `2.0.0`?

Did you try tf.contrib optimizer or tf.keras optimizer?

**Comment by SirRob1997** [comment]
Both seem to produce NaNs for me in GPU Mode. CPU works fine though

**Comment by SirRob1997** [comment]
Yes, but for my specific use case I need the AdaMax optimizer to inherit `tf.train.Optimizer` which the `tf.keras` optimizer isn't doing.

**Comment by yenlow** [comment]
I faced the same inf validation loss using LSTMs with tf 1.15 and GPUs but not with CPUs. Once I switched from Adamax to Adam, it runs great on GPUs. So perhaps, the Adamax bug is only fixed on tf 2+

**Comment by wangxu989** [comment]
BUG is in  tensorflow/core/kernel/training_ops_gpu.cc:
change :
var.device(d) -= 
        lr / (beta1_power.constant(one) - 
            beta1_power).reshape(single).broadcast(bcast)*
                 (m / (v + epsilon));
to:
var.device(d) -= 
        (lr / (beta1_power.constant(one) - 
            beta1_power)).reshape(single).broadcast(bcast)*
                 (m / (v + epsilon.reshape(single).broadcast(bcast)));
and the problem will not appear
